src/jit/optimizer.cpp

```diff
@@ -2834,365 +2829,373 @@ void Compiler::optUnrollLoops()
 
-            /* Don't unroll if too much code duplication would result. */
+        /* Don't unroll if too much code duplication would result. */
 
-            if (unrollCostSz > unrollLimitSz)
-            {
-                /* prevent this loop from being revisited */
-                optLoopTable[lnum].lpFlags |= LPFLG_DONT_UNROLL;
-                goto DONE_LOOP;
-            }
+        if (unrollCostSz > unrollLimitSz)
+        {
+            /* prevent this loop from being revisited */
+            optLoopTable[lnum].lpFlags |= LPFLG_DONT_UNROLL;
+            goto DONE_LOOP;
+        }
 
-            /* Looks like a good idea to unroll this loop, let's do it! */
-            CLANG_FORMAT_COMMENT_ANCHOR;
+        /* Looks like a good idea to unroll this loop, let's do it! */
+        CLANG_FORMAT_COMMENT_ANCHOR;
 
 #ifdef DEBUG
-            if (verbose)
+        if (verbose)
+        {
+            printf("\nUnrolling loop BB%02u", head->bbNext->bbNum);
+            if (head->bbNext->bbNum != bottom->bbNum)
             {
-                printf("\nUnrolling loop BB%02u", head->bbNext->bbNum);
-                if (head->bbNext->bbNum != bottom->bbNum)
-                {
-                    printf("..BB%02u", bottom->bbNum);
-                }
-                printf(" over V%02u from %u to %u", lvar, lbeg, llim);
-                printf(" unrollCostSz = %d\n", unrollCostSz);
-                printf("\n");
+                printf("..BB%02u", bottom->bbNum);
             }
+            printf(" over V%02u from %u to %u", lvar, lbeg, llim);
+            printf(" unrollCostSz = %d\n", unrollCostSz);
+            printf("\n");
+        }
 #endif
 
-            /* Create the unrolled loop statement list */
-
-            loopList = loopLast = nullptr;
+        /* Create the unrolled loop statement list */
+        {
+            BlockToBlockMap blockMap(getAllocator());
+            BasicBlock*     insertAfter = bottom;
 
             for (lval = lbeg; totalIter; totalIter--)
             {
-                block = head;
-
-                do
+                for (block = head->bbNext;; block = block->bbNext)
                 {
-                    GenTreeStmt* stmt;
-                    GenTree*     expr;
+                    BasicBlock* newBlock = insertAfter =
+                        fgNewBBafter(block->bbJumpKind, insertAfter, /*extendRegion*/ true);
+                    blockMap.Set(block, newBlock);
 
-                    block = block->bbNext;
-                    noway_assert(block);
-
-                    /* Visit all the statements in the block */
-
-                    for (stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
+                    if (!BasicBlock::CloneBlockState(this, newBlock, block, lvar, lval))
                     {
-                        /* Stop if we've reached the end of the loop */
+                        // cloneExpr doesn't handle everything
```

Comment:
~~Yeah, it's in `newBlock` which is set to the return from `fgNewBBafter` a few lines above.~~
No, I snip it out just below (sorry, this went through a few iterations).  I was just following suit from the [previous behavior](https://github.com/dotnet/coreclr/blob/d3ad5db7bf3209a5b0ea341cee19230d2ba0fef5/src/jit/optimizer.cpp#L3158-L3164) here.  Did I do that wrong, or do you think there's a bug before & after?  DDR passed with unroll stress... what is the ref count invariant and are we sure it's expected to hold here?  Loop unrolling [runs just before](https://github.com/dotnet/coreclr/blob/d3ad5db7bf3209a5b0ea341cee19230d2ba0fef5/src/jit/compiler.cpp#L4310-L4320) `lvaMarkLocalVars`.